builtin/consul/config_sourcer.go

```diff
@@ -0,0 +1,410 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/docs"
+	pb "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+	"github.com/mitchellh/mapstructure"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// retryInterval is the base retry value
+	retryInterval = 5 * time.Second
+
+	// maximum back off time, this is to prevent
+	// exponential runaway
+	maxBackoffTime = 180 * time.Second
+)
+
+type reqConfig struct {
+	Key        string `hcl:"key,attr"`             // kv path to retrieve
+	Namespace  string `hcl:"namespace,optional"`   // namespace the kv data resides within
+	Partition  string `hcl:"partition,optional"`   // partition the kv data resides within
+	Datacenter string `hcl:"datacenter,optional"`  // datacenter the kv data resides within
+	AllowStale bool   `hcl:"allow_stale,optional"` // whether to perform stale queries against non-leader servers
+}
+
+func (r *reqConfig) cacheKey() string {
+	return fmt.Sprintf("%s/%s/%s/%s", r.Datacenter, r.Partition, r.Namespace, r.Key)
+}
+
+type cachedVal struct {
+	kvPair *api.KVPair
+	cancel context.CancelFunc
+	err    error
+}
+
+type ConfigSourcerConfig struct {
+	// Configuration for where to talk to Consul
+	Address   string         `hcl:"address,optional"`
+	Scheme    string         `hcl:"scheme,optional"`
+	HTTPAuth  ConsulHTTPAuth `hcl:"http_auth,optional"`
+	Token     string         `hcl:"token,optional"`
+	TokenFile string         `hcl:"token_file,optional"`
+	TLSConfig TLSConfig      `hcl:"tls,optional"`
+
+	// Default location of KV data
+	Datacenter string `hcl:"datacenter,optional"`
+	Namespace  string `hcl:"namespace,optional"`
+	Partition  string `hcl:"partition,optional"`
+}
+
+func (conf *ConfigSourcerConfig) toAPIClient() (*api.Client, error) {
+	apiConfig := api.Config{
+		Address:    conf.Address,
+		Scheme:     conf.Scheme,
+		Datacenter: conf.Datacenter,
+		HttpAuth:   conf.HTTPAuth.toApiAuth(),
+		Token:      conf.Token,
+		TokenFile:  conf.TokenFile,
+		Namespace:  conf.Namespace,
+		Partition:  conf.Partition,
+		TLSConfig:  *conf.TLSConfig.toAPITLSConfig(),
```

Comment:
nit: `toAPITLSConfig()` only seems to be used in a single place, it seems strange that it returns a pointer to a struct to only then be dereferenced because the 1 place that uses it expects a struct and not a pointer. Is there a reason it's designed this way?

Reply:
Not particularly, I've updated this to:
```go
		TLSConfig: api.TLSConfig{
			Address:            conf.TLSConfig.ServerName,
			CAFile:             conf.TLSConfig.CAFile,
			CAPath:             conf.TLSConfig.CAPath,
			CAPem:              conf.TLSConfig.CAPem,
			CertFile:           conf.TLSConfig.CertFile,
			CertPEM:            conf.TLSConfig.CertPEM,
			KeyFile:            conf.TLSConfig.KeyFile,
			KeyPEM:             conf.TLSConfig.KeyPEM,
			InsecureSkipVerify: conf.TLSConfig.InsecureHTTPs,
		},
```